manpages/imodqtassist.man
.na
.nh
.TH imodqtassist 1 3.4.16 BL3DEMC
.SH NAME
imodqtassist \- Displays pages in qch file through Qt Assistant
.SH SYNOPSIS
imodqtassist  [options]  path_to_help_collection_file
.SH DESCRIPTION
This program will run Qt Assistant and send it commands to display pages
in a Qt compressed help (qch) file.  The command line arguments indicate where the
help collection file is, then an individual page is displayed when its name is
entered through standard input.  
.P
The 
.B path_to_help_collection_file
is treated as a path relative to the path in the environment variable IMOD_DIR,
unless the
.B -a
option is entered to indicate that the path is absolute.
The option
.B -p
must be entered along with the name of a 
Qt help collection (qhc) file to specify that
Qt Assistant should be started with that .qhc file. 
To provide legacy
behavior, an entry of "IMOD.adp" will be accepted and translated into "IMOD.qhc".
 The help collection file
is assumed to be relative to the 
.B path_to_help_collection_file
unless it is itself an absolute file path specification.  
.P
When a line is entered through standard input, the program attempts to display
the page.  If it succeeds, it prints 
.I Page xxx displayed
to standard error and continues waiting for more input.
If it fails to find the page, there is no error message, except in Assistant itself.
If the qhc file is not found, it issues
.I WARNING: qhc file not found.
If there is actually an error starting Qt Assistant,
there will be other error messages on the first command to display a page.  In
case of these errors, the program keeps running.
.P
Enter a line with just a "q" to terminate the program and close the current 
Qt Assistant if one is open.
.P
Each page file entry is assumed to be relative to the default prefix
.B qthelp://bl3demc/IMOD
unless the
.B -q
option is entered with a different relative or absolute prefix.  If the
.B -b
option is entered, then the prefix must be a fully qualified one for the qhc
file being used; otherwise, the prefix is simply appended to qthelp://bl3demc/IMOD/.
.P
By default, the Assistant sidebar will be started with an argument to hide the
sidebar.  Use the 
.B -k
option to omit this argument and keep the sidebar open if the user had opened
it.
.P
A completely unrelated behavior is available by running the program with
"imodqtassist -t".  In this case it will output the "Ideal thread count",
equivalent to the number of processors reported by the system, and exit.
.SH AUTHOR
David Mastronarde,  mast@colorado.edu


History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
12/6/13: Added option to Imodinfo to get contour lengths broken out by
fine-grained color, and made -o option provide object list for all outputs.

12/5/13: Updated the eTomo manpage.  Improved the eTomo help statement.
(Bug# 1764)

12/3/13: In eTomo - PEET, excluding the quoted values functionality.
(Bug# 1631)

11/30/13: Changes to Tiltalign to deal with patch tracking data: it can select
a subset of patches for the global alignment to avoid solving for too many
variables; it can do robust fitting with weights assigned either to individual
points or to whole contours.

11/28/13: Made changes in the search routine used by Tiltalign/Beadtrack to
reduce the high risk of "minimization errors" when starting close to the final
solution; also ignore particular errors after first round of robust fitting.

11/28/13: In eTomo allowing quoted values in autodocs.(Bug# 1631)

11/21/13: IMOD_4-6-33
11/21/13: Changed Corrsearch3d to use dynamic allocation, load from file more 
efficiently, and use multiple threads for some steps.

11/20/13: In eTomo - Final Aligned Stack, prevented the creation of a bad
rotation.xf file when the final aligned image stack is not present.(Bug# 1762)

11/20/13: In eTomo - Serial Sections, fixed the Newstack progress bar.
(Bug# 1749)

11/20/13: Fixed 3dmod crash from loading many files through startup window.

11/20/13: Made multi-line INFO and WARNING messages more parsable in
etomo_err.log.(Bug# 1694)

11/19/13: In eTomo - Combine, prevented the dataset from being damaged when
the Combine dialog is opened too early.(Bug# 1652)

11/19/13: In uitest, fixed a problem with finding the main panel.(Bug# 1761)

11/18/13: In eTomo made multi-line error messages parsable by adding an empty
line after each one.(Bug# 1694)

11/18/13: In eTomo - Coarse Alignment, added support for tiltxcorr parameters
SearchMagChanges and ViewsWithMagChanges.(Bug# 1760)

11/15/13: In eTomo added support for parallel processing using multiple GPUs
per computer.(Bug# 1436)

11/15/13: In eTomo reduced the excessive debug output.(Bug# 1763)

11/13/13: In eTomo - Serial Sections, added a Sobel filter checkbox and
improved the behavior of the alignment panel.(Bug# 1755)

11/11/13: IMOD_4-6-32
11/5/13: Fixed bug that crashed Imodfindbeads when adjusting size downward.

11/4/13: In eTomo - Fiduacial Model dialog, fixed a problem with deleting
autofidseed temporary files in Windows.(Bug# 1759)

11/4/13: Make xfalign apply shift limits in initial cross-correlation.

11/4/13: Fixed scale bar drawing from montage snapshot of model view window.

11/4/13: Made mrc2tif use 4 or 5 digits for all output file numbers if needed.

11/2/13: Switched display of residual model to show arrows.

11/1/13: In eTomo - Serial Sections, improved the context sensitive pop-up
menus.(Bug# 1748)

11/1/13: In eTomo - Fiducial Model, added fields for AdjustSizes and
ClusteredPointsAllowed to the Autofidseed panel.(Bug# 1743)

11/1/13: Made Tiltxcorr able to find mag changes at selected views; for a 
bidirectional series, Copytomocoms puts the option with the view number in
xcorr.com.

11/1/13: Fixed bug in imodtrans -3 option introduced 12/8/12.

10/30/13: IMOD_4-6-31
10/29/13: Fixed a problem in eTomo where the parallel panel did not appear
when eTomo reconnected to a parallel process.  This problem appeared when
Pause and Resume had been previously used.(Bug# 1757)

10/29/13: Fixed the eTomo version string.(Bug# 1744)

10/28/13: In eTomo fixed a bug where stack header data was not getting into
the dataset file when eTomo was run from batchruntomo.(Bug# 1758)

10/27/13: Added option to Tilt to get output in parallel (X/Y) slices without
inverting handedness. 

10/25/13: In eTomo removed dependencies on Cygwin.(Bug# 1713)

10/25/13: Changes in Tiltxcorr: shift limits are now applied in an elliptical
rather than rectangular region; implemented a new, more selective method for
excluding the central peak due to fixed pattern noise; added an option to
compute normalized correlation coefficients at multiple peak positions and
pick the highest (not useful so far).

10/25/13: Changed Beadtrack's elongation file output so that automatic seed
picking would not fail with only 3 beads.

10/24/13: In eTomo - Combine, removed the message complaining about tomograms
A and B not being the same size in X and Y, as this is not an error.
(Bug# 1651)

10/24/13: In eTomo, preventing embedded spaces in dataset names.(Bug# 1667)

10/24/13: Added a antialiasFilter field for the newstack panels in eTomo -
Coarse Alignment and Final Aligned Stack.(Bug# 1737)

10/23/13: In eTomo resizing when the Cleanup dialog is finished.(Bug# 1636)

10/22/13: Fixed another warping-related crash in Midas.

10/22/13: Made it possible to start Zap rubber band from ANY corner.

10/21/13: Fixed a problem that crashed 3dmod when a slicer angle is very
small, such as when doing mouse-rotating.

10/18/13: Added toolbar buttons to Zap and Slicer for drawing a single thick
arrow for snapshots; also added an object property to draw arrowheads at the
ends of contours.

10/17/13: In eTomo - Clean up, added unused, optional image files (such as the
output from mtffilter) to the file chooser.(Bug# 1509)

10/17/13: In eTomo - Clean up, added SIRT .vsr## output files.(Bug# 1533)

10/15/13: Fixed numerical problems in drawing code that crashed Midas on big
images in 64-bit Windows.

10/14/13: Added GiantCriterion, BigDiffCriterion, and ExtraLargeRadius to
eTomo - Preprocessing.(Bug# 1736)

10/9/13: Fixed a problem with running tomodataplots from eTomo under Java 1.7
that prevented the graphs from coming up.(Bug# 1754)

10/9/13: In eTomo fixed a parallel processing problem which prevented the
Resume button from working when the process was pausing.  This problem
appeared in only in reconnected processes.(Bug# 1729)

10/9/13: In eTomo - Nonlinear Anisotropic Diffusion, some test.K... file names

qttools/qtassist/imodqtassist.cpp
/*
 *  imodqtassist.cpp - Runs Qt assistant and displays pages entered through
 *                     standard input
 *
 *  Author: David Mastronarde   email: mast@colorado.edu
 *
 *  Copyright (C) 1995-2004 by Boulder Laboratory for 3-Dimensional Electron
 *  Microscopy of Cells ("BL3DEMC") and the Regents of the University of 
 *  Colorado.  See dist/COPYRIGHT for full copyright notice.
 *
 *  $Id: processchunks.cpp,v e5ab12a1256a 2011/08/23 05:58:24 sueh 
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <locale.h>
#ifndef _WIN32
#include <sys/select.h>
#include <sys/time.h>
#endif
#include <qstring.h>
#include <qapplication.h>
#include <QTimerEvent>
#include "imod_assistant.h"
#include "imodqtassist.h"

static ImodAssistant *imodHelp;

#define TIMER_INTERVAL  50
#define MAX_LINE 1024
static char threadLine[MAX_LINE];
static bool gotLine = false;
static int lineLen;

static QMutex mutex;
static AssistantThread * assThread;

static int readLine(char *line);

// MAIN
int main(int argc, char *argv[])
{
  int retval, ind = 1;
  char *qhc = NULL;
  bool absPath = false;
  bool keepBar = false;
  bool prefAbs = false;
  const char *prefix = NULL;

  if (argc == 2 && !strcmp(argv[ind], "-t")) {
    printf("Ideal thread count = %d\n", QThread::idealThreadCount());
    exit(0);
  }

  // Start the application 
  QApplication qapp(argc, argv);
  setlocale(LC_NUMERIC, "C");

  if (argc < 2) {
    fprintf(stderr, "Usage: imodqtassist [options] path_to_help_collection_file\n");
    fprintf(stderr, "   Options:\n");
    fprintf(stderr, "\t-a\tpath_to_help_collection is absolute, not relative to"
            " $IMOD_DIR\n");
    fprintf(stderr, "\t-p file\tName of qhc (help collection) file (IMOD.adp \n "
            "\t\t   can be entered instead of IMOD.qhc)\n");
    fprintf(stderr, "\t-q pref\tPrefix to use in front of help page name\n");
    fprintf(stderr, "\t-b\tPrefix is absolute, not appended to qthelp://bl3demc/IMOD/\n");
    fprintf(stderr, "\t-k\tKeep the sidebar (do not hide it)\n");
    fprintf(stderr, "\t-t\tReport ideal thread count (number of processors) "
            "and exit\n");
    exit(1);
  }

  // Parse arguments
  while (argc - ind > 1) {
    if (argv[ind][0] == '-') {
      switch (argv[ind][1]) {

      case 'a':
        absPath = true;
        break;

      case 'k':
        keepBar = true;
        break;

      case 'p':
        qhc = argv[++ind];
        break;

      case 'q':
        prefix = argv[++ind];
        break;

      case 'b':
        prefAbs = true;
        break;

      default:
        fprintf(stderr, "ERROR: unknown argument %s\n", argv[ind]);
        exit(1);
      }
    } else {
      fprintf(stderr, "ERROR: too many arguments\n");
      exit(1);
    }
    ind++;
  }

  // start the help object
  imodHelp = new ImodAssistant(argv[ind], qhc, NULL, absPath, keepBar, prefix, prefAbs);

  // Get a listener for the errors and timer
  AssistantListener *listener = new AssistantListener();
  QObject::connect(imodHelp, SIGNAL(error(const QString&)), listener,
                   SLOT(assistantError(const QString&)));

  // If using threads, start a thread to read stdin
  assThread = new AssistantThread();
  assThread->start();

  // Start timer to watch for input
  listener->startTimer(TIMER_INTERVAL);
  retval = qapp.exec();

  // Upon exit, delete the help object to close help, and kill thread if
  // needed
  delete imodHelp;

  if (assThread->isRunning())
    assThread->terminate();
  return retval;
}

void AssistantListener::assistantError(const QString &msg)
{
  fprintf(stderr, "WARNING: Qt Assistant generated error: %s\n", 
          (const char *)msg.toLatin1());
  fflush(stderr);
  //QApplication::exit(1);
}


void AssistantListener::timerEvent(QTimerEvent *e)
{
  char line[MAX_LINE];
  int err;

  // 3/23/12: Left these conditionals for historical interest
#ifdef QT_THREAD_SUPPORT

  // If there is thread support, see if the thread got a line
  // Copy the line while we have the mutex
  bool gotOne;
  mutex.lock();
  gotOne = gotLine;
  gotLine = false;
  if (gotOne)
    strcpy(line, threadLine);
  mutex.unlock();
  if (!gotOne)
    return;
#else
#ifdef _WIN32

  // If there is no thread support, Windows fallback is to process events
  // before and after.  This didn't work on second display and shouldn't
  // happen anyway
  qApp->processEvents();
#else

  // Non-windows fallback from threads uses the select operation on stdin
  // Note that it did not work to keep these as statics and get first time only
  fd_set readfds, writefds, exceptfds;
  struct timeval timeout;

  FD_ZERO(&readfds);
  FD_ZERO(&writefds);
  FD_ZERO(&exceptfds);
  FD_SET(fileno(stdin), &readfds);
  timeout.tv_sec = 0;
  timeout.tv_usec = 0;
  if (select(1, &readfds, &writefds, &exceptfds, &timeout) <= 0)
    return;
#endif

  // read line here in non-thread case
  lineLen = readLine(line);
#endif

  // Exit if blank line
  if (lineLen == 1 && threadLine[0] == 'q') {
    QApplication::exit(0);
    return;
  }
  if (!lineLen)
    return;

  // Otherwise show page and report results ( err > 0 not possible any more)
  err = imodHelp->showPage(line);
  if (err > 0)
    fprintf(stderr, "WARNING: Page %s not found\n", line);
  else
    fprintf(stderr, "Page %s displayed\n", line);
  if (err < 0 && !mWarned) {
    fprintf(stderr, "WARNING: qhc file not found\n");
    mWarned = true;
  }
    
  fflush(stderr);

#ifndef QT_THREAD_SUPPORT
#ifdef _WIN32
  qApp->processEvents();
#endif
#endif
}

// The thread loops on reading a line and sets the flag when it gets one
void AssistantThread::run()
{
  bool gotOne;
  while (1) {
    mutex.lock();
    gotOne = gotLine;
    mutex.unlock();

    // Do not go for another line until the main thread has cleared the flag
    if (gotOne) {
      usleep(20000);
      continue;
    }
    
    lineLen = readLine(threadLine);
    mutex.lock();
    gotLine = true;
    mutex.unlock();
    if (lineLen == 1 && threadLine[0] == 'q')
      return;
  }
}    

// Common routine to read a line, strip endings, and return length
static int readLine(char *line)
{
  if (!fgets(line, MAX_LINE, stdin)) {
    line[0] = 0x00;
    return 0;
  }
  if (line[MAX_LINE - 1])
    line[MAX_LINE - 1] = 0x00;
  while (strlen(line) > 0 && 
         (line[strlen(line) - 1] == '\n' || line[strlen(line) - 1] == '\r'))
    line[strlen(line) - 1] = 0x00;
  return strlen(line);
}
